pkg/apis/internal.acorn.io/v1/imageallowrules.go

```diff
@@ -0,0 +1,83 @@
+// +k8s:deepcopy-gen=package
+
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+type SignedBy struct {
+	AnyOf []string `json:"anyOf,omitempty"`
+	AllOf []string `json:"allOf,omitempty"`
+}
+
+type SignatureAnnotations struct {
+	Match       map[string]string                 `json:"match,omitempty"`
+	Expressions []metav1.LabelSelectorRequirement `json:"expressions,omitempty"`
+}
+
+func (r *SignatureAnnotations) AsSelector() (labels.Selector, error) {
+	labelselector := &metav1.LabelSelector{
+		MatchLabels:      r.Match,
+		MatchExpressions: r.Expressions,
+	}
+
+	return metav1.LabelSelectorAsSelector(labelselector)
+
+}
+
+type SignatureRules struct {
+	SignedBy    SignedBy             `json:"signedBy,omitempty"`
+	Annotations SignatureAnnotations `json:"annotations,omitempty"`
+}
+
+type ImageAllowRuleSignatures struct {
+	Sources []string         `json:"sources,omitempty"` // for now, we only support OCI registries, credentials sourced from project
+	Rules   []SignatureRules `json:"rules,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type ProjectImageAllowRulesInstance struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	Signatures ImageAllowRuleSignatures `json:"signatures,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type ProjectImageAllowRulesInstanceList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ProjectImageAllowRulesInstance `json:"items"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type ClusterImageAllowRulesInstance ProjectImageAllowRulesInstance
+
+func (in *ClusterImageAllowRulesInstance) NamespaceScoped() bool {
+	return false
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type ClusterImageAllowRulesInstanceList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ClusterImageAllowRulesInstance `json:"items"`
+}
+
+func (list *ClusterImageAllowRulesInstanceList) AsProjectImageAllowRulesInstanceList() *ProjectImageAllowRulesInstanceList {
+	var items []ProjectImageAllowRulesInstance
+	for _, item := range list.Items {
+		items = append(items, ProjectImageAllowRulesInstance(item))
+	}
+	return &ProjectImageAllowRulesInstanceList{
+		TypeMeta: list.TypeMeta,
+		ListMeta: list.ListMeta,
+		Items:    items,
+	}
```

Comment:
There is some weirdness with including `TypeMeta` and `ListMeta` here with respect to resource version, generation, and kinds.

Would returning the slice be enough? If not, then maybe leave out the `TypeMeta` and `ListMeta`?

Reply:
I see no problem returning only the slice here.